### PR TITLE
Add TestApplication to allow publish output caching

### PR DESCRIPTION
**Pull request by pakrym**
Copying files is much faster than publishing

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/TestApplication.cs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        {
+            foreach (DirectoryInfo directoryInfo in source.GetDirectories())
+            {
+                CopyFiles(directoryInfo, target.CreateSubdirectory(directoryInfo.Name), logger);
+            }
+
+            logger.LogDebug($"Processing {target.FullName}");
+            foreach (FileInfo fileInfo in source.GetFiles())
+            {
+                logger.LogDebug($"  Copying {fileInfo.Name}");
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/TestApplication.cs]: This log seem too frequent.

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/TestApplication.cs]: I thought so too. But it allowed me to see that we publish "refs" directory every time. So maybe it's not that bad?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DeploymentParameters.cs [hunk 2]
```diff
@@ -99,6 +99,8 @@ public DeploymentParameters(DeploymentParameters parameters)
             }
         }
 
+        public TestApplication TestApplication { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DeploymentParameters.cs]: This object model is circular. DeploymentParamenters has a TestApplication, but TestApplication methods take DeploymentParmameters.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/TestApplication.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/TestApplication.cs]: Maybe just print directories added?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/PublishedApplication.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        public string Path { get; }
+
+        public PublishedApplication(string path)
+        {
+            Path = path;
+        }
+
+        public void Dispose()
+        {
+            RetryHelper.RetryOperation(() => Directory.Delete(Path, true), _ => { });
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/PublishedApplication.cs]: You lost the failure log

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/PublishedApplication.cs]: Yeah, I was considering adding Dispose overload with Logger

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/PublishedApplication.cs]: Or passing the logger into the ctor.

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/PublishedApplication.cs]: That's harder in cached publisher case.

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/PublishedApplication.cs]: I lied, it's easy. Done.

**Pull request merged**